### PR TITLE
Add renderer

### LINKED ISSUE #677: Content renderer
Child of vaadin/components-team-tasks#365

See the spec: https://github.com/vaadin/components-team-tasks/issues/365#issuecomment-404149148

### PR DESCRIPTION
**Pull request by YuriyVaadin**
Fixes #677

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-combo-box/678)

### CODE REVIEW

**Comment by tomivirkki** [comment]
___
*[demo/combo-box-styling-demos.html, line 43 at r4](https://reviewable.io/reviews/vaadin/vaadin-combo-box/678#-LHbXnN14qJDZROdbJlH:-LHbXnN14qJDZROdbJlI:b3x8p1b) ([raw file](https://github.com/vaadin/vaadin-combo-box/blob/f647ff0e74937efdc276b5ad84b69f88b658b25b/demo/combo-box-styling-demos.html#L43)):*
> ```HTML
>             var comboBox = document.querySelector('#elements-renderer');
>             comboBox.items = elementsJson;
>             comboBox.renderer = (root, model) => {
> ```

Make sure the demo works with IE11 also

**Comment by tomivirkki** [comment]
___
*[src/vaadin-combo-box.html, line 223 at r3](https://reviewable.io/reviews/vaadin/vaadin-combo-box/678#-LHbPn0U4BEgCOHP59TM:-LHbY04x7VePWY1mnZm0:bq57caf) ([raw file](https://github.com/vaadin/vaadin-combo-box/blob/a836633e60dcd2191d1c82e3312dc95946dad7d3/src/vaadin-combo-box.html#L223)):*
<details><summary><i>Previously, YuriyVaadin (Yuriy Yevstihnyeyev) wrote…</i></summary><blockquote>

Updated.
</blockquote></details>

Not fixed in other files!

File: demo/combo-box-styling-demos.html [hunk 1]
```diff
@@ -55,6 +55,32 @@ <h3>Configuring Icons</h3>
...
+    </p>
+    <vaadin-demo-snippet id="combo-box-styling-demos-renderer">
+      <template preserve-content>
+        <vaadin-combo-box id="elements-renderer" label="Element" item-label-path="name" item-value-path="symbol" style="width: 240px;">
+        </vaadin-combo-box>
+        <script>
+          window.addDemoReadyListener('#combo-box-styling-demos-renderer', function(document) {
+            const comboBox = document.querySelector('#elements-renderer');
+            comboBox.items = elementsJson;
+            comboBox.renderer = function(root, model) {
```
Comment (sohrabtaee) [inline demo/combo-box-styling-demos.html]: Shouldn't it be `(root, owner, model)`?

File: test/item-renderer.html [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+      beforeEach(() => {
+        comboBox = fixture('default');
+        items = ['foo', 'bar', 'baz'];
+        comboBox.items = items;
+      });
+      afterEach(() => comboBox.opened = false);
+
+      it('should use renderer when it is defined', () => {
+        comboBox.renderer = (root, model) => {
```
Comment (sohrabtaee) [inline test/item-renderer.html]: same arguments order in here.

File: test/item-renderer.html [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+          root.appendChild(textNode);
+        };
+        comboBox.opened = true;
+
+        const item = comboBox.$.overlay._selector.querySelector('vaadin-combo-box-item');
+        expect(item.$.content.textContent.trim()).to.equal('foo');
+      });
+
+      it('renderer should receive root, model and owner', () => {
+        comboBox.renderer = (root, model, owner) => {
```
Comment (sohrabtaee) [inline test/item-renderer.html]: same arguments order in here.

File: test/item-renderer.html [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      });
+      afterEach(() => comboBox.opened = false);
+
+      it('renderer should receive empty root when defined after template', () => {
+        comboBox.renderer = root => expect(root.firstChild).to.be.null;
+        comboBox.opened = true;
+      });
+
+      it('the initial template instance should be used after renderer was assigned and removed', () => {
+        comboBox.renderer = (root, model) => {
```
Comment (sohrabtaee) [inline test/item-renderer.html]: same arguments order in here.

**Comment by tomivirkki** [comment]
___
*[src/vaadin-combo-box-mixin.html, line 57 at r8](https://reviewable.io/reviews/vaadin/vaadin-combo-box/678#-LHhVRYl3bZUB9OHGDfF:-LHhVRYl3bZUB9OHGDfG:bvzbcu9) ([raw file](https://github.com/vaadin/vaadin-combo-box/blob/5e40d71bdb35ed0c6015cd8c361359ca077a3188/src/vaadin-combo-box-mixin.html#L57)):*
> ```HTML
>          * `root` the `<vaadin-combo-box-item>` internal container element
>          *
>          * `owner` the `<vaadin-combo-box>` element
> ```

Might be vaadin-combo-box-light also (or a custom extension) so don't define the tag name here

**Comment by tomivirkki** [comment]
___
*[src/vaadin-combo-box-mixin.html, line 53 at r8](https://reviewable.io/reviews/vaadin/vaadin-combo-box/678#-LHhV6odBd7YvSMDiykk:-LHhlYk_ALmuWVpEuV4L:bxq8580) ([raw file](https://github.com/vaadin/vaadin-combo-box/blob/5e40d71bdb35ed0c6015cd8c361359ca077a3188/src/vaadin-combo-box-mixin.html#L53)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary><blockquote>

`comboBox` instead of `owner`, please
</blockquote></details>

Seems to be fixed

**Comment by tomivirkki** [comment]
___
*[src/vaadin-combo-box-item.html, line 132 at r9](https://reviewable.io/reviews/vaadin/vaadin-combo-box/678#-LHhmyhN63hjA7wc9Iyc:-LHhmyhN63hjA7wc9Iyd:bo1k3bw) ([raw file](https://github.com/vaadin/vaadin-combo-box/blob/2bbc4b63171bcca5a44bd14b2f01e143a02b1115/src/vaadin-combo-box-item.html#L132)):*
> ```HTML
>       _rendererChanged(renderer) {
>         Array.from(this.$.content.childNodes).forEach(node => {
>           this.$.content.removeChild(node);
> ```

Why is the content emptied by iterating every node? Below, in the "_ensureTemplatized" function, the content is emptied with "this.$.content.textContent = '';"

Also, it appears there are two different observer functions for the renderer. This one and "_render". Are you 100% sure they will always be called in this order? If the invocation order changed, then this function would empty any content that got rendered in the "_render" function.

**Pull request merged**